### ISSUE TITLE
Support cached feature for custom builds (Makefile)

### ISSUE THREAD

**Issue by paul-michalik** [aws/aws-sam-cli#2989]
### Describe your idea/feature/enhancement

SAM CLI should support `--cached` feature for custom builds via `Makefile`

### Proposal

Simply support the feature. Honestly I don't see technical or other reasons why this should not be possible

Things to consider:
1. Will this require any updates to the [SAM Spec](https://github.com/awslabs/serverless-application-model): No

### Additional Details

In projects with many Lambda in compiled languages (C++, C# with R2R, ...) the `build` command re-runs the build for every Lambda even without changes and valid caches.

**Comment by xazhao** [comment]
@paul-michalik Thanks for posting your idea. I already added it to our queue for internal review.

Will post here when we have any updates.

**Comment by sukruavcuoglu** [comment]
any news on this feature?

**Comment by OGoodness** [comment]
Any news?

**Comment by lvlrt** [comment]
For anybody who might be interested;

1. At the time of writing, SAM validates cache based on a hash of the directory given in CodeURI (`.` by default with `buildmethod: makefile`)
2.`buildmethod: makefile`copies the sources of the CodeURI to a (random) temporary directory where the building (aka your makefile target) will happen. e.g. in Golang this also leads to an executable with a different hash, triggering a modification in the CloudFormation deploy

Some possible solutions / ideas:
1. If you specify a CodeUri (and move the Makefile), caching will work when you do not make changes in the directory. However, if you use relative paths to shared code in a parent, and make changes there, no rebuild will trigger, and you might end up with inconsistent bugs (PS: this is also the behavior with non-custom builds, nothing changed there)
2. (specific to Go) For Go to rebuild and create consistent (hash-wise) builds, the path must be the same. This can be solved by using: `cd $(PWD); GOOS=linux GOARCH=arm64 go build -tags lambda.norpc -o $(ARTIFACTS_DIR)/bootstrap ./lambdas/your-function` where it will abandon the temporary environment set up by SAM and build in your repo. (PS: only the sources of the CodeUri are copied to the temporary environment, so this is also a way to escape that sandbox when needed)

TLDR; If you can scope the CodeUri, you can prevent cache invalidation until a more native approach is available, but checking if multiple retriggers of `sam build --cached` result in non-identic binaries `md5sum .aws-sam/build/your-function/bootstrap` is something that will prevent a lot of modifications.
